### PR TITLE
feat(sbb-link): refactor sbb-link according to new specification

### LINKED ISSUE #1239: story(sbb-link): review/refactor
### Feature Description
We want to review/refactor the lyne-button implementation.

[Figma Spec] (https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj/Lyne-Components?node-id=615%3A530767)

- dev: https://lyne.sbb.ch/components/lyne-link

**References**

- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button

**Design Spec**
- See Figma spec

**Technical Spec**
- Slot instead of label
- Should internally either use `<button>` or `<a>` (or `<span>` with visual only)
  - Should render an `<a>` element when `href` is used, when `href` is not provided use `<button>` element
  - Should support all attributes from both `<button>` or `<a>` but only use the appropriate ones
  - Should add property `static` to indicate, that internally the `<span>` should be used (To discuss: Discussion about ancester detection)
- Error should not short circuit
- Internally check if icon is used and use named slot for icons
- Refactor negative variant according to guidelines (add property negative: boolean separate from variant property)

```html
<sbb-link href="...">Action</sbb-link>
```

### Definition of Done
 - [x]  Component is implemented
 - [x]  Tests are implemented
 - [x]  E2E Tests are implemented
 - [x]  Storybook stories are implemented
 - [x]  Navigation via keyboard is tested
 - [x]  Screen reader output is tested
 - [ ]  High-contrast is tested
 - [x]  Remaining accessibility is tested
 - [x]  UX approved

### PR DESCRIPTION
**Pull request by HendrikExtSBB**
## Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/lyne-design-system/lyne-components/blob/master/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/lyne-design-system/lyne-components/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [pull request tracker](https://github.com/lyne-design-system/lyne-components/pulls) for a Pull Request (PR) that matches the one I want to submit, without success.

## Issue

This PR Closes #1239

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Changes

Changes in this pull request:
- change link text from property to slot
- add button element condition if no href is present and add corresponding attribute support
- add static-span support

## Browsers

I tested the build on the following browsers:

- [x] Firefox Desktop
- [x] Chrome Desktop
- [x] Safari Desktop
- [x] Firefox Mobile
- [x] Chrome Mobile
- [ ] Safari Mobile
- [ ] Opera
- [ ] Brave

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [x] Other (please describe):

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

- the text is now set via slot (default slot)
- the aria-label has its own property instead of the former text property

### CODE REVIEW

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:89]: We don't have a common pattern yet for internal ids. For form elements we agreed on `inputId`. Since this is not strictly speaking a form element, I would suggest to use `linkId` and use this id in all variants.
Also I would recommend initializing the id property. See e.g. https://github.com/lyne-design-system/lyne-components/blob/cc7e9a99253e64eaa3e6b479999b03d8f4fcc47f/src/components/sbb-title/sbb-title.tsx

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:53]: I'd prefer to just use `static` or was there a problem with that?

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:105]: Nitpick: I prefer methods to be placed after their usage

**Review by kyubisation — CHANGES_REQUESTED** [review]
Thank you very much for your work!
I was not able to fully finish the review, but I commented some things I noticed.

File: src/components/sbb-link/sbb-link.tsx
Comment (sebastiencloss) [inline src/components/sbb-link/sbb-link.tsx:53]: if it is initialised per default ("=false"),  no need of the "?". It is anyway optional,because it has a default initalizer.

File: src/components/sbb-link/sbb-link.tsx
Comment (sebastiencloss) [reply src/components/sbb-link/sbb-link.tsx:138]: so to add myValue, we need to add the property..
so like "value[myProperty] = myValue"  ?

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:51]:
```suggestion
   * instead of an anchor or a button
```

File: src/components/sbb-link/sbb-link.tsx
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:94]: Changed

File: src/components/sbb-link/sbb-link.tsx
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:84]: Changed

File: src/components/sbb-link/sbb-link.tsx
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:53]: Its a reserved keyword.

File: src/components/sbb-link/sbb-link.tsx
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:241]: My approach prevent delivering spans with href.

File: src/components/sbb-link/sbb-link.tsx
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:197]: I like!

**Comment by HendrikExtSBB** [comment]
DO NOT MERGE. I forgot to adjust the sbb-link-list, sorry. :D

**Review by sebastiencloss — APPROVED** [review]

**Review by Acika00mk — APPROVED** [review]

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:25]: Nitpick: This is not generally mandatory, just for the case when it `should` render as an ancher element.

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:64]: Could you also add the other accessibility properties:
```ts
/** This will be forwarded as aria-describedby to the relevant nested element. */
@Prop() public accessibilityDescribedby?: string;
/** This will be forwarded as aria-label to the relevant nested element. */
@Prop() public accessibilityLabel?: string;
/** This will be forwarded as aria-labelledby to the relevant nested element. */
@Prop() public accessibilityLabelledby?: string;
```

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:81]: This should still be variant (with `block` and `inline`)?

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:106]: As mentioned in chat; Can be placed at the bottom if you prefer it that way.

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:143]:
```suggestion
      class: this._getClassString(),
```

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:142]: If you are certain that the value cannot evaluate to falsy, you can remove the `|| undefined`

File: src/components/sbb-link/sbb-link.tsx
Comment (kyubisation) [inline src/components/sbb-link/sbb-link.tsx:165]: I'd prefer to avoid if nesting. Could we do an early return if it is static?

**Review by kyubisation — CHANGES_REQUESTED** [review]
Thanks again for your work.
Some minor issues still, after these are fixed/addressed I will merge this.

File: src/components/sbb-link/sbb-link.scss
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.scss:54]: Is line 54 really needed? Without testing I think .sbb-link--inline is sufficient here.

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [reply src/components/sbb-link/sbb-link.tsx:81]: I'd prefer to use `variant: 'block' | 'inline'` instead of a boolean

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:34]: At the moment there is a slot and a property for the icon. The property is only used to toggle showing the icon slot, so there is completely no matter what you write in the property. With the new sbb-icon component we can handle now the property and the slot. Please change the logic to use a sbb-icon inside the icon slot as a default and set the name attribute of it to the icon property. Whenever a consumer provides its own icon slot, it is overwritten and if not it takes the string from the icon property. Please also change the corresponding stories if necessary.

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [reply src/components/sbb-link/sbb-link.tsx:25]: @kyubisation Do we really want to change hrefValue into href? There are a lot of components which use hrefValue.

**Review by jeripeierSBB — COMMENTED** [review]
@kyubisation Should the logic which recognizes if a link is used inside another link or button and then becomes static be part of this PR?

File: src/components/sbb-link/sbb-link.tsx
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:25]: How do you guys want it now (@kyubisation)?^^ Sorry I don't get it fully from the comments.

Comment (jeripeierSBB) [reply src/components/sbb-link/sbb-link.tsx:25]: @HendrikExtSBB Sorry for the confusion, it should be "href" as @kyubisation previously defined and you already applied.

File: src/components/sbb-link/sbb-link.tsx
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:165]: There is no if-nesting and this condition is related to the icon position. Not if the element appears as a span.

Comment (kyubisation) [reply src/components/sbb-link/sbb-link.tsx:165]: With if-nesting I mean there is an if inside an if.
This could be refactored to:
```ts
if (this.isStatic) {
  return attributeList;
}
```

Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.tsx:165]: You were at the wrong line. ;)  But found the pos that you mentioned.

File: src/components/sbb-link/sbb-link.scss
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.scss:54]: True words, thx! :)

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:205]: Also you can remove the check for "this.icon".
```suggestion
          <slot name='icon'>
            <sbb-icon name={this.icon}></sbb-icon>
          </slot>
```

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:204]:
```suggestion
        {this.variant !== 'inline' && (
```
and the you can remove the part after the ':'

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:17]:
```suggestion
export class SbbLink implements AccessibilityProperties {
```

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:105]: Changes due to the AccessibilityProperties interface. Please rebase your PR if you don't find AccessibilityProperties interface.
```suggestion
  /** This will be forwarded as aria-label to the relevant nested element. */
  @Prop() public accessibilityLabel: string | undefined;

  /** This will be forwarded as aria-describedby to the relevant nested element. */
  @Prop() public accessibilityDescribedby: string | undefined;

  /** This will be forwarded as aria-labelledby to the relevant nested element. */
  @Prop() public accessibilityLabelledby: string | undefined;
```
```

File: src/components/sbb-link/sbb-link.scss
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.scss:85]: Is setting width and height really needed? Icon should have the right size itself.

File: src/components/sbb-link/sbb-link.scss
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.scss:92]: As reading the iconFlip docs, I think it should only be flipped on rtl mode. Why did you remove the `[dir='rtl']` part?

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:25]: Should it not be optional now (if it is a button)?
```suggestion
  @Prop() public href?: string;
```

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:62]:
```suggestion
   * Negative coloring variant flag
```

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]

File: src/components/sbb-link/sbb-link.scss
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.scss:92]: [dir='rtl'] is used to set the reading direction of text and should not be connected to to and icon that has no contextual  meaning.

Comment (jeripeierSBB) [reply src/components/sbb-link/sbb-link.scss:92]: But for what do we need the icon flip property then? Or you have at least to update the docs there. 

Your changes made that you now can flip an icon regardless of the writing direction.

**Review by jeripeierSBB — COMMENTED** [review]
General comment: I think there are more stories / usages of sbb-link inside lyne-components which have to be adapted to the new api.

File: src/components/sbb-link/sbb-link.scss
Comment (HendrikExtSBB) [reply src/components/sbb-link/sbb-link.scss]: I did not write the docs. They do not make sense to me. If you are using Icons that has a contextual meaning it should get a description somehow via an attribute (alt/aria-label or so). I dont get the idea behind that. But to flip the orientation of the icon depending on the position sounds logical to me.

File: src/components/sbb-link/sbb-link.tsx
Comment (jeripeierSBB) [inline src/components/sbb-link/sbb-link.tsx:47]: I think we have a misunderstanding regarding the iconFlip property. Can it be that you are mix the properties iconPlacement and iconFlip? Icon flip does mirror an icon which for me still only reall makes sense in rtl mode. But I am unsure if we even want to support this edge case in the sbb-link component. @kyubisation your opinion?

Comment (jeripeierSBB) [reply src/components/sbb-link/sbb-link.tsx:47]: @HendrikExtSBB  @kyubisation and I decided to remove the iconFlip property, so you can delete it and remove the corresponding css classes.

**Review by kyubisation — APPROVED** [review]
LGTM 👍 
Thank you for your work!
(FYI: I applied a few changes that I discussed yesterday with @jeripeierSBB )

**Pull request merged**
